Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -1314,6 +1318,106 @@ namespace Babylon
+            {
+                const bgfx::TextureHandle blitTextureHandle{bgfx::createTexture2D(width, height, /*hasMips*/ false, /*numLayers*/ 1, sourceTextureFormat, BGFX_TEXTURE_BLIT_DST | BGFX_TEXTURE_READ_BACK)};
+                bgfx::Encoder* encoder{GetUpdateToken().GetEncoder()};
+                encoder->blit(static_cast<uint16_t>(bgfx::getCaps()->limits.maxViews - 1), blitTextureHandle, /*dstMip*/ 0, /*dstX*/ 0, /*dstY*/ 0, /*dstZ*/ 0, sourceTextureHandle, mipLevel, x, y, /*srcZ*/ 0, width, height, /*depth*/ 0);
+
+                sourceTextureHandle = blitTextureHandle;
+                tempTexture = true;
+
+                // The requested mip level was blitted, so the source texture now has just one mip, so reset the mip level to 0.
+                mipLevel = 0;
+            }
+
+            // Allocate a buffer to store the source pixel data.
+            std::vector<uint8_t> textureBuffer(sourceTextureInfo.storageSize);
+
+            // Read the source texture.
+            m_graphicsContext.ReadTextureAsync(sourceTextureHandle, textureBuffer, mipLevel).then(arcana::inline_scheduler, *m_cancellationSource, [this, textureBuffer{std::move(textureBuffer)}, sourceTextureInfo, targetTextureInfo]() mutable {
+                // If the source texture format does not match the target texture format, convert it.
+                if (targetTextureInfo.format != sourceTextureInfo.format)
+                {
+                    std::vector<uint8_t> convertedTextureBuffer(targetTextureInfo.storageSize);
+                    bimg::imageConvert(&m_allocator, convertedTextureBuffer.data(), bimg::TextureFormat::Enum(targetTextureInfo.format), textureBuffer.data(), bimg::TextureFormat::Enum(sourceTextureInfo.format), sourceTextureInfo.width, sourceTextureInfo.height, /*depth*/ 1);
+                    textureBuffer = convertedTextureBuffer;
+                }
+
+                // Flip the image vertically if needed.
+                if (bgfx::getCaps()->originBottomLeft)
+                {
+                    FlipImage(textureBuffer, targetTextureInfo.height);
+                }
+
+                return textureBuffer;
+            }).then(m_runtimeScheduler, *m_cancellationSource, [bufferRef{Napi::Persistent(buffer)}, deferred](std::vector<uint8_t> textureBuffer) {
+                // Copy the pixel data into the JS ArrayBuffer.
+                std::memcpy(bufferRef.Value().Data(), textureBuffer.data(), textureBuffer.size());
+                deferred.Resolve(bufferRef.Value());
+            }).then(m_runtimeScheduler, arcana::cancellation::none(), [env, deferred, tempTexture, sourceTextureHandle](const arcana::expected<void, std::exception_ptr>& result) {
+                if (tempTexture)
+                {
+                    bgfx::destroy(sourceTextureHandle);
+                }
```

Comment:
This technically can still leak right now as we don't wait for continuations to complete before shutting down, but we can deal with this later after the pinnable cancellation tokens are merged.